osf/models/outcome_artifacts.py

```diff
@@ -95,3 +98,85 @@ class Meta:
             models.Index(fields=['outcome', 'artifact_type'])
         ]
         ordering = ['artifact_type', 'title']
+
+    def update_identifier(self, new_pid_value, pid_type='doi', api_request=None):
+        '''Changes the linked Identifer to one matching the new pid_value and handles callbacks.
+
+        If `finalized` is True, will also log the change on the parent Outcome if invoked via API.
+        Will attempt to delete the previous identifier to avoid orphaned entries.
+
+        Parameters:
+        new_pid_value: The string value of the new PID
+        pid_type (str): The string "type" of the new PID (for now, only "doi" is supported)
+        api_request: The api_request data from the API call that initiated the change.
+        '''
+
+        previous_identifier = self.identifier
+        self.identifier, _ = Identifier.objects.get_or_create(
+            value=new_pid_value, category=pid_type
+        )
+        self.save()
+        if previous_identifier:
+            try:
+                previous_identifier.delete()
+            except IdentifierHasReferencesError:
+                pass
+
+        if self.finalized and api_request:
+            self.outcome.log_artifact_change(
+                action=OutcomeActions.UPDATE,
+                artifact=self,
+                api_request=api_request,
+                obsolete_identifier=previous_identifier.value if previous_identifier else None,
+                new_identifier=new_pid_value
+            )
+
+    def finalize(self, api_request=None):
+        '''Sets `finalized` to True and handles callbacks.
+
+        Logs the change on the parent Outcome if invoked via the API.
+
+        Parameters:
+        api_request: The api_request data from the API call that initiated the change.
+        '''
+        if not (self.identifier and self.identifier.value and self.artifact_type):
+            return False
```

Comment:
Does this need to return `True` if it succeeds? Or is this just to bail early?

Reply:
Ack, this is from me waffling on where the validation should live. Apparently I never fully decided